File: web/src/domain/usage.test.ts
import { describe, expect, it } from "vitest";
import type { Usage } from "../api/types";
import {
  cacheRatio,
  formatCacheRatio,
  formatCost,
  formatCostByCurrency,
  formatTokenCount,
  formatUsageTokens,
  mergeUsage,
  runDisplayName,
} from "./usage";

describe("formatTokenCount", () => {
  it("renders small counts verbatim", () => {
    expect(formatTokenCount(0)).toBe("0");
    expect(formatTokenCount(999)).toBe("999");
  });

  it("renders thousands with one decimal and a k suffix", () => {
    expect(formatTokenCount(12345)).toBe("12.3k");
    expect(formatTokenCount(4100)).toBe("4.1k");
    expect(formatTokenCount(1000)).toBe("1k");
  });

  it("renders millions with an m suffix", () => {
    expect(formatTokenCount(2_500_000)).toBe("2.5m");
  });
});

describe("formatUsageTokens", () => {
  it("formats input/output as the primary, token-first figure", () => {
    const usage: Usage = {
      input_tokens: 12300,
      output_tokens: 4100,
      cached_input_tokens: 0,
      reasoning_tokens: 0,
      attempts_reported: 3,
      attempts_not_reported: 0,
    };
    expect(formatUsageTokens(usage)).toBe("12.3k in / 4.1k out");
  });
});

describe("cacheRatio / formatCacheRatio", () => {
  it("divides cached by the WHOLE prompt (input + cached), not by input alone", () => {
    // 50 cache reads beside 200 uncached input tokens is a 250-token
    // prompt, 20% of it served from cache -- not 50/200 = 25% of a prompt
    // the attempt never had.
    expect(cacheRatio({ input_tokens: 200, cached_input_tokens: 50 })).toBe(0.2);
  });

  // Task t8, claim c8: /stats rendered "588% cached" in production because
  // cache reads are reported ALONGSIDE input_tokens, never inside them, so
  // cached/input has no upper bound. A hit rate above 100% is not a fact.
  it("never exceeds 100% when cache reads dwarf the uncached input beside them", () => {
    const ratio = cacheRatio({ input_tokens: 1000, cached_input_tokens: 5880 })!;
    expect(ratio).toBeLessThanOrEqual(1);
    expect(ratio).toBeCloseTo(5880 / 6880);
    expect(formatCacheRatio(ratio)).toBe("85.5% cached");
  });

  it("reads an entirely cached prompt as 100%, not as unmeasurable", () => {
    // A fully resumed turn can report every prompt token as a cache read
    // and no uncached input at all; gating on input_tokens > 0 called that
    // unmeasurable when it is measurably 100% cached.
    expect(cacheRatio({ input_tokens: 0, cached_input_tokens: 4096 })).toBe(1);
  });

  it("never fabricates a 0/0 ratio when nothing reported any prompt tokens", () => {
    expect(cacheRatio({ input_tokens: 0, cached_input_tokens: 0 })).toBeUndefined();
  });

  it("formats a ratio as a percentage with the word 'cached'", () => {
    expect(formatCacheRatio(0.1333)).toBe("13.3% cached");
    expect(formatCacheRatio(0)).toBe("0.0% cached");
  });
});

describe("formatCost / formatCostByCurrency", () => {
  it("never invents a currency the caller did not pass", () => {
    // honesty condition h27: no currency reported means no currency shown.
    expect(formatCost(12.3)).toBe("12.30");
    expect(formatCost(12.3)).not.toContain("$");
    expect(formatCost(12.3)).not.toMatch(/[A-Z]{3}/);
  });

  it("appends the currency only when one was actually reported", () => {
    expect(formatCost(12.3, "USD")).toBe("12.30 USD");
  });

  it("formats every cost_by_currency entry independently, never summed", () => {
    expect(
      formatCostByCurrency([
        { currency: "USD", cost: 12.3 },
        { currency: "EUR", cost: 3 },
      ]),
    ).toEqual(["12.30 USD", "3.00 EUR"]);
  });
});

describe("mergeUsage", () => {
  it("sums tokens and attempt counts across entries", () => {
    const merged = mergeUsage([
      { input_tokens: 100, output_tokens: 50, cached_input_tokens: 0, reasoning_tokens: 0, attempts_reported: 1, attempts_not_reported: 0 },
      { input_tokens: 200, output_tokens: 75, cached_input_tokens: 0, reasoning_tokens: 0, attempts_reported: 1, attempts_not_reported: 1 },
    ]);
    expect(merged.input_tokens).toBe(300);
    expect(merged.output_tokens).toBe(125);
    expect(merged.attempts_reported).toBe(2);
    expect(merged.attempts_not_reported).toBe(1);
  });

  it("collapses to a single cost/currency when every entry agreed", () => {
    const merged = mergeUsage([
      { input_tokens: 1, output_tokens: 1, cached_input_tokens: 0, reasoning_tokens: 0, cost: 1, currency: "USD", attempts_reported: 1, attempts_not_reported: 0 },
      { input_tokens: 1, output_tokens: 1, cached_input_tokens: 0, reasoning_tokens: 0, cost: 2, currency: "USD", attempts_reported: 1, attempts_not_reported: 0 },
    ]);
    expect(merged.cost).toBe(3);
    expect(merged.currency).toBe("USD");
    expect(merged.cost_by_currency).toBeUndefined();
  });

  it("falls back to cost_by_currency when entries disagree on currency", () => {
    const merged = mergeUsage([
      { input_tokens: 1, output_tokens: 1, cached_input_tokens: 0, reasoning_tokens: 0, cost: 1, currency: "USD", attempts_reported: 1, attempts_not_reported: 0 },
      { input_tokens: 1, output_tokens: 1, cached_input_tokens: 0, reasoning_tokens: 0, cost: 2, currency: "EUR", attempts_reported: 1, attempts_not_reported: 0 },
    ]);
    expect(merged.cost).toBeUndefined();
    expect(merged.cost_by_currency).toEqual(
      expect.arrayContaining([
        { currency: "USD", cost: 1 },
        { currency: "EUR", cost: 2 },
      ]),
    );
  });

  it("never fabricates a cost when no entry reported one", () => {
    const merged = mergeUsage([
      { input_tokens: 1, output_tokens: 1, cached_input_tokens: 0, reasoning_tokens: 0, attempts_reported: 1, attempts_not_reported: 0 },
    ]);
    expect(merged.cost).toBeUndefined();
    expect(merged.cost_by_currency).toBeUndefined();
  });

  it("returns a present-but-empty rollup for an empty entry list", () => {
    const merged = mergeUsage([]);
    expect(merged).toEqual({
      input_tokens: 0,
      output_tokens: 0,
      cached_input_tokens: 0,
      reasoning_tokens: 0,
      attempts_reported: 0,
      attempts_not_reported: 0,
    });
    // Nothing reported any prompt tokens at all, so cache_ratio must never
    // be a fabricated 0/0 — it must be entirely absent from the merged
    // object.
    expect(merged.cache_ratio).toBeUndefined();
    expect("cache_ratio" in merged).toBe(false);
  });

  // task t2/ADR 0009: the client-side mirror of the server's rollup rules
  // for the extended usage fields.
  it("sums cached_input_tokens and reasoning_tokens across entries, mirroring the server rollup", () => {
    const merged = mergeUsage([
      { input_tokens: 400, output_tokens: 200, cached_input_tokens: 200, reasoning_tokens: 50, attempts_reported: 1, attempts_not_reported: 0 },
      { input_tokens: 2000, output_tokens: 1000, cached_input_tokens: 800, reasoning_tokens: 100, attempts_reported: 1, attempts_not_reported: 0 },
    ]);
    expect(merged.cached_input_tokens).toBe(1000);
    expect(merged.reasoning_tokens).toBe(150);
  });

  it("an entry with tokens but no cache telemetry at all contributes nothing to the cached sum, never a fabricated zero standing in for 'unmeasurable'", () => {
    const merged = mergeUsage([
      { input_tokens: 100, output_tokens: 50, cached_input_tokens: 60, reasoning_tokens: 10, attempts_reported: 1, attempts_not_reported: 0 },
      // A backend whose contract exposes no cache telemetry at all reports
      // input/output tokens with cached_input_tokens/reasoning_tokens
      // genuinely 0 (per the wire contract they're always-present sums,
      // never omitted) -- contributing 0 to the merge is exactly right.
      { input_tokens: 50, output_tokens: 20, cached_input_tokens: 0, reasoning_tokens: 0, attempts_reported: 1, attempts_not_reported: 0 },
    ]);
    expect(merged.input_tokens).toBe(150);
    expect(merged.cached_input_tokens).toBe(60);
  });

  it("recomputes cache_ratio from the merged sums rather than averaging the inputs' own ratios", () => {
    const merged = mergeUsage([
      // ratio 1/3 on its own (50 cached of a 150-token prompt)
      { input_tokens: 100, output_tokens: 0, cached_input_tokens: 50, reasoning_tokens: 0, attempts_reported: 1, attempts_not_reported: 0 },
      // ratio 0 on its own
      { input_tokens: 900, output_tokens: 0, cached_input_tokens: 0, reasoning_tokens: 0, attempts_reported: 1, attempts_not_reported: 0 },
    ]);
    // A naive average of the entries' own ratios would be 0.25; the honest
    // weighted figure over the merged sums is 50 cache reads over a
    // (1000 + 50)-token prompt.
    expect(merged.cache_ratio).toBeCloseTo(50 / 1050);
  });
});

describe("runDisplayName", () => {
  it("prefers the operator-given name", () => {
    expect(
      runDisplayName({ id: "run-1", name: "nightly build", display_hint: "hint" }),
    ).toEqual({ text: "nightly build", derived: false });
  });

  it("falls back to display_hint, marked as derived", () => {
    expect(
      runDisplayName({ id: "run-1", display_hint: "add the ledger endpoint" }),
    ).toEqual({ text: "add the ledger endpoint", derived: true });
  });

  it("falls back to the run id when neither is present, never marked as derived", () => {
    expect(runDisplayName({ id: "run-1" })).toEqual({
      text: "run-1",
      derived: false,
    });
  });
});

File: internal/api/usage_test.go
package api_test

import (
	"context"
	"encoding/json"
	"net/http"
	"testing"

	apipkg "github.com/agentculture/culture-nodes/internal/api"
	"github.com/agentculture/culture-nodes/internal/engine"
	"github.com/agentculture/culture-nodes/internal/store"
	storepg "github.com/agentculture/culture-nodes/internal/store/postgres"
)

// seedAttempt writes one attempt directly through postgres.EngineStore's
// InsertAttempt, bypassing the real claim/complete flow entirely. These
// tests are about the *shape* GET /v1alpha1/runs/{id} and GET
// /v1alpha1/node-runs render the §13.2 rollup in (task t2's acceptance
// criterion 3) — the aggregation rules themselves (retry burn included,
// unreported attempts excluded, cost never summed across currencies) are
// proven exhaustively at the store layer in
// internal/store/postgres/usage_rollup_test.go. Writing attempts directly
// here, the same way that package's own tests do, keeps this file focused
// on "does the JSON look right" without needing a retry-policy-enabled
// workflow fixture just to get two attempts onto one node run.
func seedAttempt(t *testing.T, f *fixture, nodeRunID string, n int, status engine.TechStatus, usage *engine.Usage) {
	t.Helper()
	es, err := storepg.NewEngineStore(f.store, f.nsID)
	if err != nil {
		t.Fatalf("NewEngineStore: %v", err)
	}
	err = es.InTx(context.Background(), func(ctx context.Context, tx engine.Tx) error {
		return tx.InsertAttempt(ctx, engine.Attempt{
			ID: store.NewULID(), NodeRunID: nodeRunID, Number: n, Status: status, Usage: usage,
		})
	})
	if err != nil {
		t.Fatalf("seed attempt %d on node run %s: %v", n, nodeRunID, err)
	}
}

// createMinimalRun publishes minimal.workflow.yaml and creates one run
// against it, returning the created run and its single "start" node run's
// id (the same fixture shape TestListNodeRunsAcrossRuns uses).
func createMinimalRun(t *testing.T, f *fixture) (apipkg.RunOut, string) {
	t.Helper()
	source := readFixtureWorkflow(t, "minimal.workflow.yaml")
	var published apipkg.WorkflowVersionOut
	resp, body := doJSON(t, f.client, http.MethodPost, f.url("/v1alpha1/workflows"),
		workflowSourceReq{Format: "yaml", Source: string(source)}, &published)
	requireStatus(t, resp, body, http.StatusCreated)

	var run apipkg.RunOut
	resp, body = doJSON(t, f.client, http.MethodPost, f.url("/v1alpha1/runs"),
		createRunReq{WorkflowDigest: published.Digest, Input: json.RawMessage(`{}`)}, &run)
	requireStatus(t, resp, body, http.StatusCreated)

	view := getRunView(t, f, run.ID)
	if len(view.NodeRuns) != 1 {
		t.Fatalf("run %s: got %d node runs, want 1", run.ID, len(view.NodeRuns))
	}
	return run, view.NodeRuns[0].ID
}

// TestCreateRunUsageIsPresentButEmpty proves a just-created run (zero
// attempts, let alone ones that reported usage) still carries a non-nil
// usage block on POST /v1alpha1/runs — "not yet computed" is never how a
// caller learns nothing has happened; "computed, and it says zero" is.
func TestCreateRunUsageIsPresentButEmpty(t *testing.T) {
	f := newFixture(t)
	run, _ := createMinimalRun(t, f)

	if run.Usage == nil {
		t.Fatal("Usage is nil on a freshly created run, want a present-but-empty rollup")
	}
	if run.Usage.AttemptsReported != 0 || run.Usage.AttemptsNotReported != 0 {
		t.Errorf("AttemptsReported/NotReported = %d/%d, want 0/0 (no attempts exist yet)",
			run.Usage.AttemptsReported, run.Usage.AttemptsNotReported)
	}
	if run.Usage.InputTokens != 0 || run.Usage.OutputTokens != 0 {
		t.Errorf("tokens = %d/%d, want 0/0", run.Usage.InputTokens, run.Usage.OutputTokens)
	}
	if run.Usage.Cost != nil || run.Usage.Currency != "" || len(run.Usage.CostByCurrency) != 0 {
		t.Errorf("cost fields = %+v, want all unset", run.Usage)
	}
}

// TestRunDetailUsageSumsAllAttemptsIncludingFailed is the API-level
// retry-burn shape check: GET /v1alpha1/runs/{id} must report a failed
// attempt's tokens alongside a subsequent succeeded attempt's, both summed
// into run.usage, with a single coherent currency rendered as scalar
// cost+currency.
func TestRunDetailUsageSumsAllAttemptsIncludingFailed(t *testing.T) {
	f := newFixture(t)
	run, nodeRunID := createMinimalRun(t, f)

	cost1, cost2 := 0.01, 0.02
	usd := "USD"
	seedAttempt(t, f, nodeRunID, 1, engine.StatusFailed, &engine.Usage{InputTokens: 100, OutputTokens: 40, Cost: &cost1, Currency: &usd})
	seedAttempt(t, f, nodeRunID, 2, engine.StatusSucceeded, &engine.Usage{InputTokens: 20, OutputTokens: 10, Cost: &cost2, Currency: &usd})

	view := getRunView(t, f, run.ID)
	usage := view.Run.Usage
	if usage == nil {
		t.Fatal("run detail Usage is nil")
	}
	if usage.InputTokens != 120 || usage.OutputTokens != 50 {
		t.Fatalf("tokens = %d/%d, want 120/50 (the failed attempt's usage must still count)", usage.InputTokens, usage.OutputTokens)
	}
	if usage.AttemptsReported != 2 || usage.AttemptsNotReported != 0 {
		t.Fatalf("reported/not_reported = %d/%d, want 2/0", usage.AttemptsReported, usage.AttemptsNotReported)
	}
	if usage.Cost == nil || *usage.Cost < 0.0299 || *usage.Cost > 0.0301 {
		t.Fatalf("Cost = %v, want ~0.03", usage.Cost)
	}
	if usage.Currency != "USD" {
		t.Fatalf("Currency = %q, want USD", usage.Currency)
	}
	if len(usage.CostByCurrency) != 0 {
		t.Fatalf("CostByCurrency = %+v, want empty for a single coherent currency", usage.CostByCurrency)
	}
}

// TestRunDetailUsageDistinguishesUnreportedFromZero proves the API shape
// keeps "nobody reported usage" (attempts_reported: 0) visibly distinct
// from a token sum that merely happens to equal zero.
func TestRunDetailUsageDistinguishesUnreportedFromZero(t *testing.T) {
	f := newFixture(t)
	run, nodeRunID := createMinimalRun(t, f)

	seedAttempt(t, f, nodeRunID, 1, engine.StatusFailed, nil)
	seedAttempt(t, f, nodeRunID, 2, engine.StatusCancelled, nil)

	view := getRunView(t, f, run.ID)
	usage := view.Run.Usage
	if usage == nil {
		t.Fatal("run detail Usage is nil")
	}
	if usage.AttemptsReported != 0 || usage.AttemptsNotReported != 2 {
		t.Fatalf("reported/not_reported = %d/%d, want 0/2", usage.AttemptsReported, usage.AttemptsNotReported)
	}
	if usage.InputTokens != 0 || usage.OutputTokens != 0 {
		t.Fatalf("tokens = %d/%d, want 0/0", usage.InputTokens, usage.OutputTokens)
	}
}

// TestRunDetailUsageMixedCurrenciesExposedSeparately proves a run whose
// attempts priced their work in two different currencies never collapses
// into one summed number: the API renders cost_by_currency (a list) and
// leaves the scalar cost/currency fields unset.
func TestRunDetailUsageMixedCurrenciesExposedSeparately(t *testing.T) {
	f := newFixture(t)
	run, nodeRunID := createMinimalRun(t, f)

	usd, jpy := "USD", "JPY"
	costUSD, costJPY := 1.0, 100.0
	seedAttempt(t, f, nodeRunID, 1, engine.StatusSucceeded, &engine.Usage{InputTokens: 10, OutputTokens: 5, Cost: &costUSD, Currency: &usd})
	seedAttempt(t, f, nodeRunID, 2, engine.StatusSucceeded, &engine.Usage{InputTokens: 10, OutputTokens: 5, Cost: &costJPY, Currency: &jpy})

	view := getRunView(t, f, run.ID)
	usage := view.Run.Usage
	if usage == nil {
		t.Fatal("run detail Usage is nil")
	}
	if usage.Cost != nil {
		t.Fatalf("Cost = %v, want unset (mixed currencies must never be summed into one scalar)", *usage.Cost)
	}
	if usage.Currency != "" {
		t.Fatalf("Currency = %q, want empty when mixed", usage.Currency)
	}
	if len(usage.CostByCurrency) != 2 {
		t.Fatalf("CostByCurrency = %+v, want two entries", usage.CostByCurrency)
	}
	byCurrency := map[string]float64{}
	for _, cc := range usage.CostByCurrency {
		byCurrency[cc.Currency] = cc.Cost
	}
	if byCurrency["USD"] != 1.0 || byCurrency["JPY"] != 100.0 {
		t.Fatalf("byCurrency = %+v, want USD=1 JPY=100", byCurrency)
	}
}

// TestRunDetailUsageCacheRatioComputedHonestly proves GET
// /v1alpha1/runs/{id} renders task t2/ADR 0009's cached_input_tokens,
// reasoning_tokens, and a computed cache_ratio (cached/input, only when
// input_tokens > 0) — and that an attempt reporting tokens with no cache
// telemetry at all contributes nothing to the cached sum, never a
// fabricated zero standing in for "unmeasurable".
func TestRunDetailUsageCacheRatioComputedHonestly(t *testing.T) {
	f := newFixture(t)
	run, nodeRunID := createMinimalRun(t, f)

	cachedTokens, reasoningTokens := int64(9984), int64(500)
	seedAttempt(t, f, nodeRunID, 1, engine.StatusSucceeded,
		&engine.Usage{InputTokens: 13880, OutputTokens: 200, CachedInputTokens: &cachedTokens, ReasoningTokens: &reasoningTokens})
	// A second attempt reports usage but no cache telemetry (e.g. a backend
	// whose contract exposes none) -- must not pollute the sum with a zero.
	seedAttempt(t, f, nodeRunID, 2, engine.StatusSucceeded,
		&engine.Usage{InputTokens: 100, OutputTokens: 20})

	view := getRunView(t, f, run.ID)
	usage := view.Run.Usage
	if usage == nil {
		t.Fatal("run detail Usage is nil")
	}
	if usage.InputTokens != 13980 {
		t.Fatalf("InputTokens = %d, want 13980", usage.InputTokens)
	}
	if usage.CachedInputTokens != 9984 {
		t.Fatalf("CachedInputTokens = %d, want 9984 (only attempt 1 reported any)", usage.CachedInputTokens)
	}
	if usage.ReasoningTokens != 500 {
		t.Fatalf("ReasoningTokens = %d, want 500", usage.ReasoningTokens)
	}
	if usage.CacheRatio == nil {
		t.Fatal("CacheRatio is nil, want computed (input+cached > 0)")
	}
	// Issue #47's real codex numbers: 9984 cache reads beside 13880 uncached
	// input tokens. The whole prompt is 23964 tokens, so 41.7% of it was
	// served from cache -- 9984/13980 would claim 71.4% of a prompt that
	// never existed.
	wantRatio := 9984.0 / 23964.0
	if *usage.CacheRatio < wantRatio-0.0001 || *usage.CacheRatio > wantRatio+0.0001 {
		t.Fatalf("CacheRatio = %v, want ~%v", *usage.CacheRatio, wantRatio)
	}
}

// TestRunDetailUsageCacheRatioOmittedWhenNoInputTokensReported proves
// cache_ratio never renders as a fabricated 0 when no attempt in scope
// reported any prompt tokens at all -- the boundary case a naive
// division would get wrong (0/0).
func TestRunDetailUsageCacheRatioOmittedWhenNoInputTokensReported(t *testing.T) {
	f := newFixture(t)
	run, _ := createMinimalRun(t, f)

	view := getRunView(t, f, run.ID)
	usage := view.Run.Usage
	if usage == nil {
		t.Fatal("run detail Usage is nil")
	}
	if usage.CacheRatio != nil {
		t.Fatalf("CacheRatio = %v, want nil (no attempt reported any prompt tokens)", *usage.CacheRatio)
	}
	if usage.CachedInputTokens != 0 || usage.ReasoningTokens != 0 {
		t.Fatalf("cached/reasoning tokens = %d/%d, want 0/0 (sum of nothing)", usage.CachedInputTokens, usage.ReasoningTokens)
	}
}

// TestNodeRunsListingCarriesPerNodeRunUsage proves GET /v1alpha1/node-runs
// (the cross-run listing) carries each row's own usage rollup: a
// dispatched-and-completed node run's usage reflects what that attempt
// reported, and a never-dispatched node run reports zero attempts either
// way (reported and not_reported both 0), not a missing/null usage block.
func TestNodeRunsListingCarriesPerNodeRunUsage(t *testing.T) {
	f := newFixture(t)
	source := readFixtureWorkflow(t, "minimal.workflow.yaml")
	var published apipkg.WorkflowVersionOut
	resp, body := doJSON(t, f.client, http.MethodPost, f.url("/v1alpha1/workflows"),
		workflowSourceReq{Format: "yaml", Source: string(source)}, &published)
	requireStatus(t, resp, body, http.StatusCreated)

	var dispatchedRun, undispatchedRun apipkg.RunOut
	resp, body = doJSON(t, f.client, http.MethodPost, f.url("/v1alpha1/runs"),
		createRunReq{WorkflowDigest: published.Digest, Input: json.RawMessage(`{}`)}, &dispatchedRun)
	requireStatus(t, resp, body, http.StatusCreated)
	resp, body = doJSON(t, f.client, http.MethodPost, f.url("/v1alpha1/runs"),
		createRunReq{WorkflowDigest: published.Digest, Input: json.RawMessage(`{}`)}, &undispatchedRun)
	requireStatus(t, resp, body, http.StatusCreated)

	dispatchedNodeRunID := getRunView(t, f, dispatchedRun.ID).NodeRuns[0].ID
	undispatchedNodeRunID := getRunView(t, f, undispatchedRun.ID).NodeRuns[0].ID

	actorID := f.insertActor("worker")
	claimed := f.claim("worker-1", dispatchedNodeRunID)
	cost := 0.5
	usd := "USD"
	if _, err := f.api.Engine.CompleteAttempt(context.Background(), engine.CompletionRequest{
		WorkID:       claimed.ID,
		WorkerID:     "worker-1",
		FencingToken: claimed.FencingToken,
		Attempt:      int(claimed.Attempt),
		TechStatus:   engine.StatusSucceeded,
		Outcome:      "completed",
		Output:       json.RawMessage(`{}`),
		ActorID:      actorID,
		Usage:        &engine.Usage{InputTokens: 42, OutputTokens: 8, Cost: &cost, Currency: &usd},
	}); err != nil {
		t.Fatalf("complete node run %s: %v", dispatchedNodeRunID, err)
	}

	var list apipkg.NodeRunListOut
	resp, body = doJSON(t, f.client, http.MethodGet, f.url("/v1alpha1/node-runs?limit=50"), nil, &list)
	requireStatus(t, resp, body, http.StatusOK)

	byID := make(map[string]apipkg.NodeRunListItemOut)
	for _, item := range list.Items {
		byID[item.ID] = item
	}

	dispatched, ok := byID[dispatchedNodeRunID]
	if !ok {
		t.Fatalf("dispatched node run %s not found in listing", dispatchedNodeRunID)
	}
	if dispatched.Usage == nil {
		t.Fatal("dispatched node run's Usage is nil")
	}
	if dispatched.Usage.InputTokens != 42 || dispatched.Usage.OutputTokens != 8 {
		t.Fatalf("dispatched tokens = %d/%d, want 42/8", dispatched.Usage.InputTokens, dispatched.Usage.OutputTokens)
	}
	if dispatched.Usage.AttemptsReported != 1 || dispatched.Usage.AttemptsNotReported != 0 {
		t.Fatalf("dispatched reported/not_reported = %d/%d, want 1/0", dispatched.Usage.AttemptsReported, dispatched.Usage.AttemptsNotReported)
	}
	if dispatched.Usage.Cost == nil || *dispatched.Usage.Cost != 0.5 || dispatched.Usage.Currency != "USD" {
		t.Fatalf("dispatched cost/currency = %v/%q, want 0.5/USD", dispatched.Usage.Cost, dispatched.Usage.Currency)
	}

	undispatched, ok := byID[undispatchedNodeRunID]
	if !ok {
		t.Fatalf("undispatched node run %s not found in listing", undispatchedNodeRunID)
	}
	if undispatched.Usage == nil {
		t.Fatal("undispatched node run's Usage is nil (want a present-but-empty rollup, not absent)")
	}
	if undispatched.Usage.AttemptsReported != 0 || undispatched.Usage.AttemptsNotReported != 0 {
		t.Fatalf("undispatched reported/not_reported = %d/%d, want 0/0 (no attempts exist at all)",
			undispatched.Usage.AttemptsReported, undispatched.Usage.AttemptsNotReported)
	}
}

// TestRunsListDoesNotComputeUsage locks in the documented scope boundary:
// GET /v1alpha1/runs (the bulk listing) does not pay for a per-row rollup
// query, unlike run detail and the node-runs listing. Its RunOut.Usage
// stays nil/omitted, distinguishable on the wire from a computed-but-empty
// rollup by the key's absence entirely.
func TestRunsListDoesNotComputeUsage(t *testing.T) {
	f := newFixture(t)
	_, _ = createMinimalRun(t, f)

	var list apipkg.RunListOut
	resp, body := doJSON(t, f.client, http.MethodGet, f.url("/v1alpha1/runs"), nil, &list)
	requireStatus(t, resp, body, http.StatusOK)
	if len(list.Items) == 0 {
		t.Fatal("expected at least one run in the list")
	}
	for _, r := range list.Items {
		if r.Usage != nil {
			t.Fatalf("run %s: Usage = %+v, want nil on the bulk listing", r.ID, r.Usage)
		}
	}
}
